### ISSUE TITLE
[BUG] fraud_detection never release record for every user and every number

### ISSUE THREAD

**Issue by okhowang** [OpenSIPS/opensips#3411]
**OpenSIPS version you are running**

```
```

**Describe the bug**

**To Reproduce**

**Expected behavior**

**Relevant System Logs**

**OS/environment information**
 - Operating System: 
 - OpenSIPS installation: 
 - other relevant information:

**Additional context**

**Comment by okhowang** [comment]
And matched_len may not be initialized when number has char except 0~9
It may cause unexpected behaviour.
https://github.com/OpenSIPS/opensips/blob/master/modules/fraud_detection/fraud_detection.c#L281

**Comment by bogdan-iancu** [comment]
Could you detail a bit more on the leak you found? like what are the leaking structure and the case when the leak happens?

**Comment by okhowang** [comment]
`fraud_detection` save every user's statistic in memory forever (util shutdown).
if there are 1,000,000 user in a system. it will has 1,000,000 record kept in memory.
even there only 1000 user call in one day.
maybe we should clear statistic every day/week.


I found this problem because, my service OOM.
and I notice `map_get` use a lot of mem. which is called by `fraud_detection`. and a lot of entries cost 64k memory.
it because uninitialized `matched_len` which make normal user's string become a huge random string.

uninitialized `matched_len` and unclearable statistic cause the OOM.

**Comment by liviuchircu** [comment]
Thank you for the investigation work and your patch, @okhowang!  I applied a slightly better fix and closed #3413, could you pull it and check if the "leak" persists?  While I am fully aware that the fraud-detection stores `(user, prefix)` tuples, it is not clear as to what degree they will fill up the SHM pool.  Questions such as the following could help us get a better idea for SHM pool sizing in your case, or maybe even develop a "stat timer" feature for the module:

A) is the calling user pool finite (e.g. from `subscriber` table) or random (e.g. simply use `$fU`)?
B) does the majority of users typically dial random numbers frequently (calling center), or is it a residential service?

Cheers!

**Comment by okhowang** [comment]
> I applied a slightly better fix and closed https://github.com/OpenSIPS/opensips/pull/3413, could you pull it and check if the "leak" persists?

thank you. But I'm still at 3.2 branch. I will upgrade to 3.4 LTS later. So, feedback may be delayed for weeks.
But I think it should work as expected.

> A) is the calling user pool finite (e.g. from subscriber table) or random (e.g. simply use $fU)?

It is finite. We use `subscriber` table currently.

> B) does the majority of users typically dial random numbers frequently (calling center), or is it a residential service?

Yes, we provide a calling center service, but also as a residential service.


I haven't calculated the exact memory usage for every user and rule/prefix. I just think it should be clear in period. 
10,000 users maybe use 100MB.
of course, the uninitialized match_len make the problem exacerbated.

**Comment by liviuchircu** [comment]
> of course, the uninitialized match_len make the problem exacerbated.

Exactly, this was my initial thought as well.  Let's see if the situation improves after the fix, you could easily cherry-pick it into your `3.2` source code and make a new build.  Certainly, there are ways to shoot yourself in the foot with OpenSIPS when it comes to memory usage (for example, if you do `cache_store()` with `$ci` (Call-ID) as insert key, this is 100% a memory leak.  Does `check_fraud()` suffer from such an "algorithmic" memory leak too?  Remains to be seen/proven.

**Comment by okhowang** [comment]
`show_fraud_stats` can only show one user and one prefix now.
I found how many data store in memory by gdb a coredump file.
maybe we can make `show_fraud_stats` dump all data if called without parameters.
it may be more simple to observe the relation between map entries and memory usage.
